FFI adjustments (SR-6142)

compnerd opened the pull request:
<!-- What's in this pull request? -->
Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

compnerd committed:
IRGen: minor clean up to explosion handling (NFC)

Indicate that the type is const for the cast.  Use `Range` and a
for-loop rather than the explicit counters.  NFC.

compnerd committed:
IRGen: follow target specific ABI for FFI

When lowering the arguments for the FFI call on certain targets, LLVM
expects that the parameter be passed as a structure rather than the
extracted values.  Enhance the argument lowering for the call and the
function construction to ensure that the type can be flattened according
to the ABI for the target before attempting to explode it for the SIL
IRGen lowering.  This allows building for targets such as Linux ARM HF
and Linux ARM SF target with the HF floating point ABI.

Addresses SR-6142!

compnerd commented:
@swift-ci please test

aschwaighofer commented on lib/IRGen/GenCall.cpp:
```diff
@@ -938,13 +938,16 @@ llvm::Type *SignatureExpansion::expandExternalSignatureTypes() {
+      // If the coercion type is a struct which can be flattened, we need to
+      // expand it.
+      auto *ArgTy = AI.getCoerceToType();
+      if (AI.isDirect() && AI.getCanBeFlattened() &&
```
IRGen uses lower-case convention: ArgTy -> argTy

AI.isDirect() is not necessary here. This code is guarded by  'case clang::CodeGen::ABIArgInfo::Direct'.

compnerd replied:
Sure on the case thing, but the `AI.isDirect()` is needed as on L924, there is a `LLVM_FALLTHROUGH` for `clang::CodeGen::ABIArgInfo::Extend`.

aschwaighofer replied:
Oh, I missed that.

aschwaighofer commented on lib/IRGen/GenCall.cpp:
```diff
@@ -1684,9 +1687,10 @@ static void emitCoerceAndExpand(IRGenFunction &IGF,
-                                       Explosion &out) {
+static void emitDirectExternalArgument(IRGenFunction &IGF,
+                                       const clang::CodeGen::ABIArgInfo &AI,
+                                       SILType argType, llvm::Type *ArgTy,
```
ArgTy -> argTy

aschwaighofer commented on lib/IRGen/GenCall.cpp:
```diff
@@ -1696,7 +1700,8 @@ static void emitDirectExternalArgument(IRGenFunction &IGF, SILType argType,
   // Check to see if we can pairwise coerce Swift's exploded scalars
   // to Clang's expanded elements.
-  if (canCoerceToSchema(IGF.IGM, expandedTys, inputSchema)) {
+  if (AI.getCanBeFlattened() &&
```
Do we need a check for AI.isDirect() here?

This can be called from a AI.isExtend() context.

compnerd replied:
Oh, I missed that, I'll add that.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
*sigh* there was another hidden `LLVM_FALLTHROUGH` that missed both of our eyes.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

rjmccall commented on lib/IRGen/GenCall.cpp:
```diff
@@ -1880,7 +1885,7 @@ static void externalizeArguments(IRGenFunction &IGF, const Callee &callee,
       }
 
-      emitDirectExternalArgument(IGF, paramType, toTy, in, out);
+      emitDirectExternalArgument(IGF, AI, paramType, toTy, in, out);
```
Passing toTy is now redundant.

compnerd replied:
Removed.

rjmccall commented on lib/IRGen/GenCall.cpp:
```diff
@@ -938,13 +938,16 @@ llvm::Type *SignatureExpansion::expandExternalSignatureTypes() {
-          ParamIRTypes.push_back(expandedType->getElementType(j));
+      // If the coercion type is a struct which can be flattened, we need to
+      // expand it.
+      auto *ArgTy = AI.getCoerceToType();
```
'type' was fine as the name of this variable, and this code would probably look somewhat cleaner if you just dyn_cast'ed to StructType outside of the condition.

compnerd replied:
I was trying to use similar names in the various places.  Renamed it to `coercedTy`.

rjmccall commented on lib/IRGen/GenCall.cpp:
```diff
@@ -1684,19 +1687,21 @@ static void emitCoerceAndExpand(IRGenFunction &IGF,
-                                       Explosion &out) {
+static void emitDirectExternalArgument(IRGenFunction &IGF,
+                                       const clang::CodeGen::ABIArgInfo &AI,
+                                       SILType argType, llvm::Type *argTy,
```
The rename to 'argTy' seems strictly worse (for the local variable you should introduce instead of passing this separately).

Also, please just replace toTy in the parameter order; "source before target" is a nice thing to try to be consistent about.

compnerd replied:
Replaced with a local `coercedTy` variable.  Sure, I can move the `ABIArgInfo` in the place of `toTy`, but that seems like additional information rather than source/target to me.

rjmccall replied:
It's the real "to type" information here: we don't need to produce a value of some abstract IR type, we need to produce a value matching the argument type scheme described by AI.

rjmccall commented on lib/IRGen/GenCall.cpp:
```diff
@@ -1961,23 +1966,24 @@ bool irgen::addNativeArgument(IRGenFunction &IGF, Explosion &in,
-                                       llvm::Type *coercionTy,
-                                       Explosion &out,
+                                       const clang::CodeGen::ABIArgInfo &AI,
+                                       Explosion &in, Explosion &out,
```
Same comment about reordering.  The order was "source value, source foreign info, target value, target type, target type info", and now it's rather muddled.

compnerd replied:
Swapped the `AI` for the `coercionTy`.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test

compnerd commented:
i386 iOS seems to mismatch now too; adjusted.  Is there an easy way to get the darwin coverage?

compnerd commented:
@swift-ci please test macOS platform

compnerd commented:
@swift-ci please test macOS platform

compnerd commented:
@swift-ci please test macOS platform

compnerd commented:
The duplication of the checks in `abitypes.swift` is rather annoying, especially as they are subtly different :/

compnerd commented:
@swift-ci please test macOS platform

compnerd commented:
@swift-ci please test macOS platform

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test macOS platform

compnerd commented:
@rjmccall @aschwaighofer I think that this is ready now

rjmccall commented on test/IRGen/abitypes.swift:
```diff
@@ -420,19 +435,28 @@ class Foo {
   // x86_64-ios: [[SEL:%[0-9]+]] = load i8*, i8** @"\01L_selector(invert:)", align 8
-  // x86_64-ios: [[NEG:%[0-9]+]] = call zeroext i1 bitcast (void ()* @objc_msgSend to i1 (%1*, i8*, i1)*)(%1* [[RECEIVER:%[0-9]+]], i8* [[SEL]], i1 zeroext %0)
+  // x86_64-ios: store i1 %0, i1* %coerced-arg.coerced._value, align 1
+  // x86_64-ios: [[CAST:%[0-9]+]] = bitcast %TSb* %coerced-arg.coerced to i1*
```
Can these references to ``%coerced-arg.coerced`` reasonably be a ``[[VARIABLE]]`` reference instead?

compnerd replied:
Not sure about the x86_64 macosx case, but the others ones are regular enough to do.

rjmccall commented on test/IRGen/unexploded-calls.swift:
```diff
@@ -0,0 +1,32 @@
+// CHECK:   store float %1, float* %coerced-arg.coerced.g._value, align 4
+// CHECK:   %3 = bitcast %TSC1SV* %coerced-arg.coerced to %struct.S*
+// CHECK:   %4 = load %struct.S, %struct.S* %3, align 4
+// CHECK:   call void @f(%struct.S %4)
```
Same thing: we prefer to use FileCheck variables when the original LLVM value name seems fairly arbitrary.

compnerd committed:
test: use more FileCheck patterns (NFC)

Try to use more FileCheck pattern matching rather than hardcoding
variable names.

compnerd commented:
@rjmccall there, I think that Ive converted over to file check patterns as best as I could, if there is a better way, I can adjust that.

compnerd commented:
@swift-ci please test

aschwaighofer commented on lib/IRGen/GenCall.cpp:
```diff
@@ -1685,18 +1688,21 @@ static void emitCoerceAndExpand(IRGenFunction &IGF,
 
   // Check to see if we can pairwise coerce Swift's exploded scalars
   // to Clang's expanded elements.
-  if (canCoerceToSchema(IGF.IGM, expandedTys, inputSchema)) {
```
Sorry for noticing this so late but this is not NFC for darwin.

We can see this in the tests. They should not change (but there are added extra loads/stores of i8*)

I believe what is going wrong is that this check should be something like (indirect arguments should be coercible):
if (
 (AI.direct() && AI.getCanBeFlattened()) || 
  !AI.direct()) &&
canCoerceToSchema(...))

compnerd replied:
Hmm, I thought that @rjmccall was okay with the additional coercion (since it should get optimized out).  In any case, I'll see if that works for avoiding the unnecessary memory traffic.

rjmccall replied:
Well, if we can avoid coercing through memory, great.

compnerd replied:
Adjusted to avoid the coercion through memory unless needed.  I expect that the linux build should be fine (it passed locally).

compnerd committed:
IRGen: avoid unnecessary coercion memory traffic

Attempt to avoid unnecessary traffic for the type coercion unless it is
really needed.

compnerd commented:
@swift-ci please test

compnerd commented:
@swift-ci please test Linux platform

compnerd commented:
@aschwaighofer addressed your comments

aschwaighofer commented on lib/IRGen/GenCall.cpp:
```diff
@@ -1960,24 +1969,24 @@ bool irgen::addNativeArgument(IRGenFunction &IGF, Explosion &in,
-  // the storage type of a type is an llvm::StructType or it has a
-  // single-element explosion.
+  ArrayRef<llvm::Type*> expandedTys;
+  if (AI.isDirect() && AI.getCanBeFlattened() &&
```
The same isDirectFlattened || isIndirect logic should apply here too. We are also calling this for indirect parameters. Or am I mistaken?

This should make any changes in the existing tests unneccessary. I believe with the change as is lowering of indirect (eg. i8*) parameters at function entry still goes through memory. The optimizer would optimize this away but at Onone this is unecessary overhead.

compnerd replied:
Isn't that incorrect though?  I thought that we don't want to flatten the parameter unless we are passing it directly.

aschwaighofer replied:
Oh sorry, I was getting confused here this is only checking for the case where we have to expand. An indirect type would also never be an llvm::StructType.

You are right of course.

slavapestov commented on test/IRGen/unexploded-calls.swift:
```diff
@@ -0,0 +1,32 @@
+// RUN: %swift -target thumbv7-unknown-windows-msvc -parse-stdlib -parse-as-library -I %S/Inputs/usr/include -module-name Swift -S -emit-ir -o - %s | %FileCheck %s
```
@compnerd This test is failing in PR testing

compnerd replied:
Interesting.  Wonder how it passed during this run.  Do we not have the ARM backend enabled?  The warning originates from clang (its not a swift side thing).  I'll look into this.
